app/src/main/java/com/example/ncov19traking/data/NCoVRepository.kt

```diff
@@ -1,86 +1,117 @@
-            nCoVDao.loadYesterday().first()
+            when (val response =
+                ApiResponse.create(NCoVApiClient.nCoVApi.getYesterdayGeneralNumbers())) {
+                is ApiSuccessResponse -> {
+                    nCoVDao.saveYesterday(response.data)
+                    nCoVDao.loadYesterday().first()
+                }
+                is ApiSuccessEmptyResponse -> nCoVDao.loadYesterday().first()
+                is ApiErrorResponse -> {
+                    Log.d("apiError", "${response.code}: ${response.message}")
+                    nCoVDao.loadYesterday().first()
+                }
+            }
         } catch (e: Exception) {
-            if (nCoVDao.loadYesterday().size != 0)
+            if (nCoVDao.loadYesterday().isNotEmpty())
                 nCoVDao.loadYesterday().first()
             else
                 NCoVInfoYesterday(0, 0, 0, 0)
         }
     }
 
-    private suspend fun refreshAllYesterdayCases() {
-        nCoVDao.saveYesterday(NCoVApiAdapter.nCoVApi.getYesterdayGeneralNumbers())
-    }
-
     suspend fun getAllCountries(): Array<NumbersByCountry>? {
         return try {
-            refreshAllCountries()
-            countryDao.load()
+            when (val response = ApiResponse.create(NCoVApiClient.nCoVApi.getNumbersByCountry())) {
+                is ApiSuccessResponse -> {
+                    countryDao.save(response.data)
+                    countryDao.load()
+                }
+                is ApiSuccessEmptyResponse -> countryDao.load()
+                is ApiErrorResponse -> {
+                    Log.d("apiError", "${response.code}: ${response.message}")
+                    errorResponse.postValue(ErrorBody(response.code, response.message))
+                    countryDao.load()
+                }
+            }
         } catch (e: IOException) {
             Log.d("debug", "${e.cause}: ${e.message}")
+            errorResponse.postValue(e.message?.let { ErrorBody(message = it) })
             if (countryDao.load().isNotEmpty())
                 countryDao.load()
             else emptyArray()
         }
     }
 
-    private suspend fun refreshAllCountries() {
-        countryDao.save(NCoVApiAdapter.nCoVApi.getNumbersByCountry())
 
-    }
-
-    suspend fun getHistoricalCountryData() = NCoVApiAdapter.nCoVApi.getHistoricalDataByCountry()
+    suspend fun getHistoricalCountryData() = NCoVApiClient.nCoVApi.getHistoricalDataByCountry()
 
     suspend fun getAllHistoricalDataCases(): Timeline? {
         return try {
-            refreshAllHistoricalDataCases()
-            globalHistoricalDao.load().first()
-        }catch (e : Exception){
-            if (globalHistoricalDao.load().size != 0)
+            when (val response =
+                ApiResponse.create(NCoVApiClient.nCoVApi.getAllHistoricalData())) {
+                is ApiSuccessResponse -> {
+                    globalHistoricalDao.save(response.data)
+                    globalHistoricalDao.load().first()
```

Comment:
repeated code

Reply:
would you mind give me more information here?

Reply:
if you are going to call `globalHistoricalDao.load()` in every `when` branch even in the `catch` in the case of an exception why not writing that piece of code only once at the beginning or end of the method, wherever you think is best.

Reply:
check it now